server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -846,7 +899,21 @@ public Query termQuery(Object value, SearchExecutionContext context) {
 
         @Override
         protected void parseCreateField(DocumentParserContext context) {
-            // field defined as runtime field, don't index anything
+            // Run-time fields are mapped to this mapper, so it needs to handle storing values for use in synthetic source.
+            // #parseValue calls this method once the run-time field is created.
+            if (context.dynamic() == ObjectMapper.Dynamic.RUNTIME && context.canAddIgnoredField()) {
```

Comment:
It is not clear to me which code path leads to this and does not have source already stored at `DocumentParser` level.  E.g. if `canAddIgnoredField` is true we would have already stored source inside `parseObjectDynamic` right after creating this mapper.

Reply:
Took me a while to capture this.. When a new runtime field is created, we go through the path in `parseDynamicValue` in the master. Then, in the data node, the forwarded mapping contains the runtime field so this code executes; `parseValue` calls `parseObjectOrField` with this mapper, instead of `parseDynamicValue`.

Reply:
Thanks.